src/source/promises.spec.bs

```diff
@@ -38,294 +38,671 @@ namespace tests
-		sub _()
-			promises.chain(promises.allSettled(invalid)).then(sub(result)
-				m.testSuite.fail("should not get here")
-			end sub).catch(sub(error)
-				m.testSuite.assetEqual(error.message, "Did not supply an array")
-				m.testSuite.assetNotInvalid(error.backtrace)
-			end sub).finally(sub()
-				m.testSuite.done()
-			end sub)
-		end sub
+		function _()
+			context = {
+				thenCount: 0
+				catchCount: 0
+				errorMessage: ""
+				backtrace: invalid
+			}
+
+			results = promises.chain(promises.allSettled(invalid), context).then(sub(_, context)
+				context.thenCount++
+			end sub).catch(function(error, context)
+				context.catchCount++
+				context.errorMessage = error.message
+				context.backtrace = error.backtrace
+			end function).toPromise()
+
+			return promises.onFinally(results, sub(context)
+				backtrace = context.backtrace
+				context.delete("backtrace")
+				m.testSuite.assertEqual(context, {
+					thenCount: 0
+					catchCount: 1
+					errorMessage: "Did not supply an array"
+				})
+				m.testSuite.assertNotInvalid(backtrace)
+			end sub, context)
+		end function
 
-		@async
 		@it("handled empty array")
-		sub _()
-			promises.chain(promises.allSettled([])).then(sub(result)
-				m.testSuite.assertEqual(result, [])
-			end sub).catch(sub(error)
-				m.testSuite.fail("should not get here")
-			end sub).finally(sub()
-				m.testSuite.done()
-			end sub)
-		end sub
+		function _()
+			context = {
+				thenCount: 0
+				catchCount: 0
+				result: invalid
+			}
+
+			results = promises.chain(promises.allSettled([]), context).then(sub(result, context)
+				context.thenCount++
+				context.result = result
+			end sub).catch(sub(_, context)
+				context.catchCount++
+			end sub).toPromise()
+
+			return promises.onFinally(results, sub(context)
+				m.testSuite.assertEqual(context, {
+					thenCount: 1
+					catchCount: 0
+					result: []
+				})
+			end sub, context)
+		end function
 
-		@async
 		@it("resolving all promises in allSettled")
-		sub _()
-			promises.chain(promises.allSettled([
+		function _()
+			context = {
+				thenCount: 0
+				catchCount: 0
+				result: invalid
+			}
+
+			results = promises.chain(promises.allSettled([
 				promises.resolve(1)
 				promises.resolve(2)
 				promises.resolve(3)
-			])).then(sub(_)
-				msg = "resolved promise result should be [1,2,3]"
-				m.testSuite.assertTrue(rooibos.common.eqArray(_, [
-					{ status: promises.PromiseState.resolved, value: 1 },
-					{ status: promises.PromiseState.resolved, value: 2 },
-					{ status: promises.PromiseState.resolved, value: 3 }
-				]), msg)
-			end sub).catch(sub(_)
-				m.testSuite.fail("should not get here")
-			end sub).finally(sub()
-				m.testSuite.done()
-			end sub)
-		end sub
+			]), context).then(sub(result, context)
+				context.thenCount++
+				context.result = result
+			end sub).catch(sub(_, context)
+				context.catchCount++
+			end sub).toPromise()
+
+			return promises.onFinally(results, sub(context)
```

Comment:
Why not utilize the .finally() on the end of each chain above? This feels like extra code for no added benefit.  (applies to all the tests in this file)

Reply:
you prefer the way they read. that's fine...